File: ts/test/entity/story_link/StoryLinkEntity.test.ts

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'


import { ShortcutSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveDelay,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


describe('StoryLinkEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when SHORTCUT_TEST_LIVE=TRUE.
  afterEach(liveDelay('SHORTCUT_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = ShortcutSDK.test()
    const ent = testsdk.StoryLink()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.SHORTCUT_TEST_LIVE
    for (const op of ['create', 'update', 'load', 'remove']) {
      if (maybeSkipControl(t, 'entityOp', 'story_link.' + op, live)) return
    }

    const setup = basicSetup()
    // The basic flow consumes synthetic IDs and field values from the
    // fixture (entity TestData.json). Those don't exist on the live API.
    // Skip live runs unless the user provided a real ENTID env override.
    if (setup.syntheticOnly) {
      t.skip('live entity test uses synthetic IDs from fixture — set SHORTCUT_TEST_STORY_LINK_ENTID JSON to run live')
      return
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select


    // CREATE
    const story_link_ref01_ent = client.StoryLink()
    let story_link_ref01_data = setup.data.new.story_link['story_link_ref01']

    story_link_ref01_data = await story_link_ref01_ent.create(story_link_ref01_data)
    assert(null != story_link_ref01_data.id)


    // UPDATE
    const story_link_ref01_data_up0: any = {}
    story_link_ref01_data_up0.id = story_link_ref01_data.id

    const story_link_ref01_markdef_up0 = { name: 'created_at', value: 'Mark01-story_link_ref01_' + setup.now }
    ;(story_link_ref01_data_up0 as any)[story_link_ref01_markdef_up0.name] = story_link_ref01_markdef_up0.value

    const story_link_ref01_resdata_up0 = await story_link_ref01_ent.update(story_link_ref01_data_up0)
    assert(story_link_ref01_resdata_up0.id === story_link_ref01_data_up0.id)

    assert((story_link_ref01_resdata_up0 as any)[story_link_ref01_markdef_up0.name] === story_link_ref01_markdef_up0.value)


    // LOAD
    const story_link_ref01_match_dt0: any = {}
    story_link_ref01_match_dt0.id = story_link_ref01_data.id
    const story_link_ref01_data_dt0 = await story_link_ref01_ent.load(story_link_ref01_match_dt0)
    assert(story_link_ref01_data_dt0.id === story_link_ref01_data.id)


    // REMOVE
    const story_link_ref01_match_rm0: any = { id: story_link_ref01_data.id }
    await story_link_ref01_ent.remove(story_link_ref01_match_rm0)
  

  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/story_link/StoryLinkTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = ShortcutSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['story_link01','story_link02','story_link03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  // Detect whether the user provided a real ENTID JSON via env var. The
  // basic flow consumes synthetic IDs from the fixture file; without an
  // override those synthetic IDs reach the live API and 4xx. Surface this
  // to the test so it can skip rather than fail.
  const idmapEnvVal = process.env['SHORTCUT_TEST_STORY_LINK_ENTID']
  const idmapOverridden = null != idmapEnvVal && idmapEnvVal.trim().startsWith('{')

  const env = envOverride({
    'SHORTCUT_TEST_STORY_LINK_ENTID': idmap,
    'SHORTCUT_TEST_LIVE': 'FALSE',
    'SHORTCUT_TEST_EXPLAIN': 'FALSE',
    'SHORTCUT_APIKEY': 'NONE',
  })

  idmap = env['SHORTCUT_TEST_STORY_LINK_ENTID']

  const live = 'TRUE' === env.SHORTCUT_TEST_LIVE

  if (live) {
    client = new ShortcutSDK(merge([
      {
        apikey: env.SHORTCUT_APIKEY,
      },
      extra
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.SHORTCUT_TEST_EXPLAIN,
    live,
    syntheticOnly: live && !idmapOverridden,
    now: Date.now(),
  }

  return setup
}
  

File: go/test/story_link_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/shortcut-sdk/go"
	"github.com/voxgig-sdk/shortcut-sdk/go/core"

	vs "github.com/voxgig-sdk/shortcut-sdk/go/utility/struct"
)

func TestStoryLinkEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.StoryLink(nil)
		if ent == nil {
			t.Fatal("expected non-nil StoryLinkEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := story_linkBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "story_link." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set SHORTCUT_TEST_STORY_LINK_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		storyLinkRef01Ent := client.StoryLink(nil)
		storyLinkRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "story_link"}, setup.data), "story_link_ref01"))

		storyLinkRef01DataResult, err := storyLinkRef01Ent.Create(storyLinkRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		storyLinkRef01Data = core.ToMapAny(storyLinkRef01DataResult)
		if storyLinkRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if storyLinkRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// UPDATE
		storyLinkRef01DataUp0Up := map[string]any{
			"id": storyLinkRef01Data["id"],
		}

		storyLinkRef01MarkdefUp0Name := "created_at"
		storyLinkRef01MarkdefUp0Value := fmt.Sprintf("Mark01-story_link_ref01_%d", setup.now)
		storyLinkRef01DataUp0Up[storyLinkRef01MarkdefUp0Name] = storyLinkRef01MarkdefUp0Value

		storyLinkRef01ResdataUp0Result, err := storyLinkRef01Ent.Update(storyLinkRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		storyLinkRef01ResdataUp0 := core.ToMapAny(storyLinkRef01ResdataUp0Result)
		if storyLinkRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if storyLinkRef01ResdataUp0["id"] != storyLinkRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if storyLinkRef01ResdataUp0[storyLinkRef01MarkdefUp0Name] != storyLinkRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", storyLinkRef01MarkdefUp0Name, storyLinkRef01ResdataUp0[storyLinkRef01MarkdefUp0Name])
		}

		// LOAD
		storyLinkRef01MatchDt0 := map[string]any{
			"id": storyLinkRef01Data["id"],
		}
		storyLinkRef01DataDt0Loaded, err := storyLinkRef01Ent.Load(storyLinkRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		storyLinkRef01DataDt0LoadResult := core.ToMapAny(storyLinkRef01DataDt0Loaded)
		if storyLinkRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if storyLinkRef01DataDt0LoadResult["id"] != storyLinkRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		storyLinkRef01MatchRm0 := map[string]any{
			"id": storyLinkRef01Data["id"],
		}
		_, err = storyLinkRef01Ent.Remove(storyLinkRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

	})
}

func story_linkBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "story_link", "StoryLinkTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read story_link test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse story_link test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"story_link01", "story_link02", "story_link03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("SHORTCUT_TEST_STORY_LINK_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"SHORTCUT_TEST_STORY_LINK_ENTID": idmap,
		"SHORTCUT_TEST_LIVE":      "FALSE",
		"SHORTCUT_TEST_EXPLAIN":   "FALSE",
		"SHORTCUT_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["SHORTCUT_TEST_STORY_LINK_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["SHORTCUT_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["SHORTCUT_APIKEY"],
			},
			extra,
		})
		client = sdk.NewShortcutSDK(core.ToMapAny(mergedOpts))
	}

	live := env["SHORTCUT_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["SHORTCUT_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
